### PR TITLE
Use LTV ratio instead of collateralization

### PR DESCRIPTION
**Pull request by bonomat**
Using LTV ratios instead of collateralization in percent is easier to calculate with due to the simple formula of :
`LTV = repayment_amount / (collateral * rate)`
With this formula alone we can compute if the user provided enough collateral or at what rate the loan will be liquidated given a max LTV ratio.

Signed-off-by: Philipp Hoenisch <philipp@hoenisch.at>

### CODE REVIEW

File: bobtimus/src/loan.rs [hunk 1]
```diff
@@ -334,19 +329,29 @@ enum LoanValidationError {
...
+    #[error(
+        "The requested ltv {requested_ltv} was not offered (available ltvs: {:?})",
+        offered_ltvs
+    )]
+    InvalidRequestedLtv {
+        requested_ltv: Decimal,
+        offered_ltvs: Vec<Decimal>,
+    },
+    #[error(
+        "The provided ltv {provided_ltv} does not match the not selected LTV {requested_ltv})"
```
Comment (da-kami) [inline bobtimus/src/loan.rs:350]:
```suggestion
        "The provided ltv {provided_ltv} does not match the selected LTV {requested_ltv})"
```

File: bobtimus/src/loan.rs [hunk 2]
```diff
@@ -531,17 +557,18 @@ mod tests {
                 days: 30,
                 interest_mod: Decimal::ZERO,
             }],
-            collateralizations: vec![],
+            // not relevant for test but to show as reference - that's what the borrower can chose
+            initial_ltvs: vec![Ltv { ltv: dec!(0.70) }],
 
             // irrelevant for this test
```
Comment (da-kami) [inline bobtimus/src/loan.rs:563]: I suppose this comment should be removed given that you fill in values?

File: bobtimus/src/loan.rs [hunk 3]
```diff
@@ -562,7 +589,7 @@ mod tests {
         // The `liquidation_price` is calculated by :
         // `(repayment_amount / current_ltv) / collateral_amount`
         // or with numbers in this example:
-        // (10_500 / 0.75 ) / 1.5 = 9333.33333333
+        // (10000 / 0.75 ) / 1.5 = 9333.33333333
```
Comment (da-kami) [inline bobtimus/src/loan.rs:592]: Why is the repayment amount `10000` now, if the `principal` and `interest_rate` did not change? 
That cannot be correct - with a `principal` of `10,000` and an interest rate of `5%` the `repayment_amount` should still be `10,500` - or am I missing something?

File: waves/src/Bobtimus.tsx [hunk 4]
```diff
@@ -56,14 +51,14 @@ export interface LoanOffer {
...
-    collateralizations: Collateralization[];
+    initial_ltvs: Ltv[];
 }
 
 export interface LoanRequest {
     principal_amount: number;
     collateral_amount: number;
     collateral_inputs: { txin: OutPoint; original_txout: any; blinding_key: string }[];
-    collateralization: number;
+    ltv: number;
```
Comment (da-kami) [inline waves/src/Bobtimus.tsx:61]: Why does the borrower send an LTV? 
The current LTV of the borrower can be calculated from `=REPAYMENT_AMOUNT/(LOAN_REQUEST_COLLTERAL*CURRENT_RATE)`. This feels weird, because the borrower now sends values that are used to calculate the LTV, but at the same time provides an LTV. Is this just to cross check that the calculation on UI and Bobtimus is the same?

File: e2e_tests/e2e_test_setup.sh [hunk 5]
```diff
@@ -12,7 +12,7 @@ while getopts "CS" o; do
             then
                 echo "Clearing docker compose data"
                 cd nigiri
-                docker-compose down
+                docker compose down
```
Comment (da-kami) [inline e2e_tests/e2e_test_setup.sh:15]: nit: These changes (if committed on purpose) would be nice in a separate commit. I don't see a relation to the code changes.

File: bobtimus/src/lib.rs [hunk 6]
```diff
@@ -304,16 +304,7 @@ where
...
-                Collateralization {
-                    collateralization: dec!(1.5),
-                    interest_mod: Decimal::ZERO,
-                },
-                Collateralization {
-                    collateralization: dec!(2.0),
-                    interest_mod: Decimal::ZERO,
-                },
-            ],
+            initial_ltvs: vec![Ltv { ltv: dec!(0.5) }, Ltv { ltv: dec!(0.7) }],
```
Comment (da-kami) [inline bobtimus/src/lib.rs:307]: I am not sure about the the name `initial_ltv` - are these thresholds for the borrower?

File: bobtimus/src/loan.rs [hunk 7]
```diff
@@ -51,8 +52,8 @@ pub struct LoanOffer {
     /// Interest in relation to terms
     pub terms: Vec<Term>,
 
-    /// Interest rates in relation to collteralization
-    pub collateralizations: Vec<Collateralization>,
+    /// Interest rates in relation to ltv
```
Comment (da-kami) [inline bobtimus/src/loan.rs:55]: The `Ltv` struct does not represent a relation to the interest rate. I think this comment is incorrect.

**Comment by da-kami** [comment]
An overall note on price vs. rate:

In my previous PR I started introducing price when talking about the exchange rate. It seems that this is not helping, because now we started mixing `price` and `rate` in code and comments.

How about we remove the `price` again, but rename the `rate` to `exchange_rate` so it is clear we are talking about the exchange rate and not the interest rate in calculations?

File: bobtimus/src/loan.rs [hunk 8]
```diff
@@ -62,12 +63,11 @@ pub struct Term {
...
 #[derive(Debug, Clone, Copy, serde::Serialize)]
-pub struct Collateralization {
-    pub collateralization: Decimal,
-    /// Interest to be added on top of the base interest rate for this term.
-    pub interest_mod: Decimal,
+pub struct Ltv {
+    pub ltv: Decimal,
 }
 
 // TODO: Make sure that removing sat_per_vbyte is OK here
```
Comment (da-kami) [inline bobtimus/src/loan.rs:73]: sidenote: this comment is outdated and can be removed (sry, I forgot to remove it)

File: bobtimus/src/loan.rs [hunk 9]
```diff
@@ -110,24 +113,29 @@ pub fn loan_calculation_and_validation(
...
     current_price: LiquidUsdt,
 ) -> Result<ValidatedLoan> {
-    let interest_rate = calculate_interest_rate(
-        loan_request.term,
-        loan_request.collateralization,
-        &loan_offer.terms,
-        &loan_offer.collateralizations,
-        loan_offer.base_interest_rate,
-    )?;
+    let selected_term = match loan_offer
```
Comment (da-kami) [inline bobtimus/src/loan.rs:116]: This is validated double now - there is still a check in `validate_loan_is_acceptable` for the term. 
My idea was to keep **all** the validation steps in one place and accept that there are some calculations that we might do beforehand even though the loan is not acceptable. 

With introducing this early exit scenario we have to maintain this validation step twice in code. 
Either we remove it here again, or we remove it in the validation step. If we remove the validation step we split up the validation again and don't contain it in one function.

File: bobtimus/src/lib.rs (hunk 6 above)
Comment (da-kami) [reply bobtimus/src/lib.rs:307]: OK, given the code in `validate_loan_is_acceptable` this is not an interval, but we check against exact matches.

Most user interfaces I looked at don't let the user chose the LTV but the collateralization rate - and then calculate the LTV. I think with this change the user will have to select the LTV. I think we can do that, just pointing out the implication.

File: bobtimus/src/loan.rs [hunk 10]
```diff
@@ -403,27 +408,34 @@ fn validate_loan_is_acceptable(
...
-            .collateralization;
-
-        if request_collateralization < min_collateralization {
-            return Ok(Err(LoanValidationError::CollateralizationBelowMin {
-                request_collateralization,
-                min_collateralization,
-            }));
-        }
+    // we allow for slight price variations of 1% up or down
+    let one_percent = requested_ltv
```
Comment (da-kami) [inline bobtimus/src/loan.rs:412]: With this we have two price fluctuation intervals now - one for the actual accepted eschange rate and one for the  LTV.

File: bobtimus/src/loan.rs [hunk 11]
```diff
@@ -809,41 +829,79 @@ mod tests {
...
+            .with_provided_ltv(provided_ltv)
+            .with_ltvs(ltvs);
+
+        validate_loan_is_acceptable(loan_validation_params)
+            .unwrap()
+            .unwrap();
+    }
+
+    #[test]
+    fn given_loan_request_with_in_list_is_ok() {
```
Comment (da-kami) [inline bobtimus/src/loan.rs:896]:
```suggestion
    fn given_loan_request_within_list_is_ok() {
```

File: bobtimus/src/loan.rs [hunk 12]
```diff
@@ -809,41 +829,79 @@ mod tests {
...
+        assert_eq!(
+            error,
+            LoanValidationError::InvalidProvidedLtv {
+                requested_ltv,
+                provided_ltv
+            }
+        );
+    }
+    #[test]
+    fn requested_ltv_is_slightly_different_to_provided_ltv() {
```
Comment (da-kami) [inline bobtimus/src/loan.rs:880]:
```suggestion
    fn given_requested_ltv_is_slightly_different_to_provided_ltv_then_ok() {
```

File: bobtimus/src/loan.rs (hunk 2 above)
Comment (bonomat) [reply bobtimus/src/loan.rs:563]: It's still irrelevant but  I want to have the values there nevertheless :)

**Comment by da-kami** [comment]
Other things I noticed that we should clean up before merging:

* I think the `max_ltv` is not needed anymore because this check should already be covered by finding the `requested_ltv` in the `initial_ltvs`

File: e2e_tests/e2e_test_setup.sh (hunk 5 above)
Comment (bonomat) [reply e2e_tests/e2e_test_setup.sh:15]: wasn't on purpose...
